### PR TITLE
Kevin-Cotrina_Testing-Nerdery-Challenge

### PR DESCRIPTION
**Pull request by kcotrinam**
### **In this branch I:**

- Added tests for all the methods.
- Completed 100% of coverage.

### CODE REVIEW

File: src/index.spec.ts [hunk 1]
```diff
@@ -4,9 +4,172 @@ import {
...
-  createProduct 
+  createProduct,
 } from './index';
 
-test('Testing time!', () => {
-  console.log('welcome');
+import fetch from 'node-fetch';
+import { mocked } from 'ts-jest/utils';
+
+describe('Test isInteger function', () => {
```
Comment (ederivero) [inline src/index.spec.ts:13]: for more readability when you define a describe block only put the name of the function to be tested, for this case should be `describe('isInteger',...)`

File: src/index.spec.ts [hunk 2]
```diff
@@ -4,9 +4,172 @@ import {
...
+  it('return false if we pass a string', () => {
+    expect(isInteger('1')).toBe(false);
+  });
+
+  it('return false if we pass a NaN', () => {
+    expect(isInteger(NaN)).toBe(false);
+  });
+
+  it('return false if we pass a null', () => {
+    expect(isInteger(null)).toBe(false);
```
Comment (ederivero) [inline src/index.spec.ts:35]: all the `it` functions should start with the word 'should' for more readability

File: src/index.spec.ts [hunk 3]
```diff
@@ -4,9 +4,172 @@ import {
...
+  });
+});
+
+describe('Test toLowerCase function', () => {
```
Comment (ederivero) [inline src/index.spec.ts:39]: `describe('toLowerCase ',()=>{`

File: src/index.spec.ts [hunk 4]
```diff
@@ -4,9 +4,172 @@ import {
...
+  it('returns lowercase string', () => {
+    expect(toLowerCase('Hello')).toBe('hello');
+  });
+
+  it('returns lowercase string', () => {
+    expect(toLowerCase('Hello World')).toBe('hello world');
+  });
+
+  it("return 'Please provide a string' if we pass an empty string", () => {
```
Comment (ederivero) [inline src/index.spec.ts:48]: `it('should...',()=>{...})`

File: src/index.spec.ts [hunk 5]
```diff
@@ -4,9 +4,172 @@ import {
...
+    const arr = [1, 'test', 23, 1, 'test', 23];
+
+    expect(removeDuplicatesFromArray(arr)).toEqual([1, 'test', 23]);
+  });
+
+  it('should fail if we pass a wrong argument type', () => {
+    const prop = { number: 2 };
+    const functionTest = () => removeDuplicatesFromArray(prop as any);
+
+    expect(functionTest).toThrowError(Error);
```
Comment (ederivero) [inline src/index.spec.ts:74]: Only `Error` is too generic and you are not validating the message, you could put `toThrowError(new Error('please provide an array of numbers or strings')` and doing this, you are validating the error and his message

File: src/index.spec.ts [hunk 6]
```diff
@@ -4,9 +4,172 @@ import {
...
+  it('returns lowercase string', () => {
+    expect(toLowerCase('Hello World')).toBe('hello world');
+  });
+
+  it("return 'Please provide a string' if we pass an empty string", () => {
+    expect(toLowerCase('')).toBe('Please provide a string');
+  });
+});
+
+describe('Test removeDuplicatesFromArray function', () => {
```
Comment (ederivero) [inline src/index.spec.ts:53]: `describe('removeDuplicatesFromArray',()=>{`

File: src/index.spec.ts [hunk 7]
```diff
@@ -4,9 +4,172 @@ import {
...
+
+  it('should fail if we pass a wrong argument type', () => {
+    const prop = { number: 2 };
+    const functionTest = () => removeDuplicatesFromArray(prop as any);
+
+    expect(functionTest).toThrowError(Error);
+  });
+});
+
+describe('Test createProduct function', () => {
```
Comment (ederivero) [inline src/index.spec.ts:78]: `describe('createProduct', ()=>{`

File: src/index.spec.ts [hunk 8]
```diff
@@ -4,9 +4,172 @@ import {
...
+  const validProduct = {
+    name: 'Product',
+    tags: ['tag1'],
+    description: 'description product 1',
+    price: 10.2,
+  };
+
+  const invalidProduct = {
+    ...validProduct,
+    tags: ['tag1', 'tag2'],
```
Comment (ederivero) [inline src/index.spec.ts:88]: You could use `tags: [...validProduct.tags, 'tag2']` instead of replacing

File: src/index.spec.ts [hunk 9]
```diff
@@ -4,9 +4,172 @@ import {
...
+      price: 10.2,
+    });
+  });
+
+  it('should fail if we pass an ivnalid schema of the product', () => {
+    expect(() => createProduct(invalidProduct)).toThrowError(Error);
+  });
+});
+
+describe('Test createRandomProduct function', () => {
```
Comment (ederivero) [inline src/index.spec.ts:106]: `describe('createRandomProduct', () => {`

File: src/index.spec.ts [hunk 10]
```diff
@@ -4,9 +4,172 @@ import {
...
+      name: expect.any(String),
+      tags: expect.any(Array),
+      description: expect.any(String),
+      price: expect.any(String),
+    });
+  });
+
+  it('should throw an error if we pass a user without <creator> role', () => {
+    expect(() => createRandomProduct('diana@themyscira.com')).toThrowError(
+      Error,
```
Comment (ederivero) [inline src/index.spec.ts:119]: `new Error('You are not allowed to create products')`

File: src/index.spec.ts [hunk 11]
```diff
@@ -4,9 +4,172 @@ import {
...
+  };
+
+  const invalidProduct = {
+    ...validProduct,
+    tags: ['tag1', 'tag2'],
+  };
+
+  it('should return a product with an id', () => {
+    expect(createProduct(validProduct)).toEqual({
+      id: expect.any(Number),
```
Comment (ederivero) [inline src/index.spec.ts:93]: If you are only validating if the result has an id you could use `toHaveProperty('id')` or `expect(createProduct(validProduct).id).toBeDefined();`

File: src/index.spec.ts [hunk 12]
```diff
@@ -4,9 +4,172 @@ import {
...
+      name: 'Product',
+      tags: ['tag1'],
+      description: 'description product 1',
+      price: 10.2,
+    });
+  });
+
+  it('should fail if we pass an ivnalid schema of the product', () => {
```
Comment (ederivero) [inline src/index.spec.ts:101]: `should throw an Error if pass an invalid product schema`

File: src/index.spec.ts [hunk 13]
```diff
@@ -4,9 +4,172 @@ import {
...
+    expect(() => createProduct(invalidProduct)).toThrowError(Error);
```
Comment (ederivero) [inline src/index.spec.ts:102]: `expect(() => createProduct(invalidProduct)).toThrowError(
      new Error(
        '[{"message":"\\"tags\\" must contain 1 items","path":["tags"],"type":"array.length","context":{"limit":1,"value":["tag1","tag2"],"label":"tags","key":"tags"}}]',
      ),
    );`

File: src/index.spec.ts [hunk 14]
```diff
@@ -4,9 +4,172 @@ import {
...
+
+jest.mock('node-fetch', () => {
+  return jest.fn();
+});
+
+beforeEach(() => {
+  mocked(fetch).mockClear();
+});
+
+describe('test getStarWarsPlanets function ', () => {
```
Comment (ederivero) [inline src/index.spec.ts:132]: `describe('getStarWarsPlanets', () => {`

File: src/index.spec.ts [hunk 15]
```diff
@@ -4,9 +4,172 @@ import {
...
+    tags: ['tag1', 'tag2'],
+  };
+
+  it('should return a product with an id', () => {
+    expect(createProduct(validProduct)).toEqual({
+      id: expect.any(Number),
+      name: 'Product',
+      tags: ['tag1'],
+      description: 'description product 1',
+      price: 10.2,
```
Comment (FrankFuentes21) [inline src/index.spec.ts:97]: You can create the object on a separate constant and then pass it to the toEqual method

File: src/index.spec.ts [hunk 16]
```diff
@@ -4,9 +4,172 @@ import {
...
+});
+
+describe('Test createRandomProduct function', () => {
+  it('should create a product with the correct role', () => {
+    expect(createRandomProduct('clark@kent.com')).toEqual({
+      id: expect.any(Number),
+      name: expect.any(String),
+      tags: expect.any(Array),
+      description: expect.any(String),
+      price: expect.any(String),
```
Comment (FrankFuentes21) [inline src/index.spec.ts:113]: we can use the toHaveProperty here, also, is better to define the object in a separated constant and then pass it to the methods to evaluate the structure

File: src/index.spec.ts [hunk 17]
```diff
@@ -4,9 +4,172 @@ import {
...
+  });
+});
+
+jest.mock('node-fetch', () => {
+  return jest.fn();
+});
+
+beforeEach(() => {
+  mocked(fetch).mockClear();
+});
```
Comment (FrankFuentes21) [inline src/index.spec.ts:130]: could be a good idea having this at the beginning of the file

**Pull request closed without merging**